### DISCUSSION TITLE
How to boot Windows over PCIe NVME?

### DISCUSSION THREAD

**Discussion by alkisg** [CloverHackyColor/CloverBootloader#352] [Q&A] [answered]
Hi! I'm completely new to Clover and I'm not sure if it fits my needs:
Without involving macOS/Hackintosh at all, is it possible to use Clover in order to boot Windows in the following scenario?

- A Samsung M.2 disk
- An [M.2 to PCIe adapter card](https://www.amazon.co.uk/Akasa-AK-PCCM2P-01-Adapter-Profile-Support/dp/B01LZMIBVP)
- An old Dell UEFI bios that doesn't see the PCIe adapter

I can put Clover in any media, e.g. USB stick, SD card, second old SATA HDD. But:
- Will NvmExpressDxe.efi actually allow the Windows bootmgfw.efi to load WinLoad.efi? Does NvmExpressDxe.efi somehow insert itself in the UEFI services list, so that the Windows loader can use it?
- I assume the boot order is UEFI > Clover > bootmgfw.efi > WinLoad.efi, right? No BIOS/CSM involved...
- How can I configure Clover? Do I need to manually create a config.plist?

So far, I downloaded https://github.com/CloverHackyColor/CloverBootloader/releases/download/5129/CloverV2-5129.zip and I wrote it in a fat32 USB stick, and it booted, but it just showed an Options etc menu, with no operating system entries, and the "S" key doesn't open a shell, so I'm not sure what I'm supposed to do there.

In another try without Clover, I put the EFI partition in an old SATA HDD, and the Windows partition in the NVME drive, but Windows failed to boot with a blue screen saying that [winload.efi is missing](https://www.diskpart.com/screenshot/en/others/windows-10/winload-efi-missing.png).

Thank you for any feedback!

**Comment by alkisg** [comment]
Ah when I tried with a SATA disk in UEFI/GPT, Clover did manage to see it and automatically create an entry.
But when I cloned it to M.2, it didn't see it.

I guess Clover just can't see my PCIe card where my M.2 is attached.

**Comment by alkisg** [comment, marked as answer]
I managed to boot it!

Quick instructions, as it really took a lot of experimentation:

- Run http://cvad-mac.narod.ru/index/bootdiskutility_exe/0-5 in order to prepare the boot USB / SD / CF card.
- Remember to copy NvmExpressDxe.efi from off to UEFI/BIOS subdirs.
- Boot in **BIOS/GPT** mode, not in UEFI/GPT!
- Clover emulates an UEFI environment, and it's able to load NvmExpressDxe.efi in its own environment, but not in the motherboard's UEFI environment. That was the part that I couldn't even imagine!
- The boot entries will be autodetected but not saved. In order to get the Timeout=5 to work, you need to select the boot entry inside Clover, press space to see its UUID, then boot to Windows, edit config.plist, and set DefaultVolume to the correct UUID.

**Reply by glnzglnz** [reply]
alkisg - trying to do what you did, but a few Qs if you have the time:

1.   http://cvad-mac.narod.ru/index/bootdiskutility_exe/0-5 doesn't work any more.  What to do?
2.   How do I download Clover from its github page and burn it onto a USB stick?  I'm a complete noob and don't know my way around github at all.  The numerous github source files are completely impenetrable to me - would much appreciate a step-by-step.  The github page I found is at [https://github.com/CloverHackyColor/CloverBootloader](url)
3.   So I need to go into the BIOS settings of my Optiplex 7010 and return it to classic BIOS, not UEFI?  That is really non-intuitive.  Did you ever set yours back to UEFI?
4.   How do I find the UUIDs in your last bullet point?
5.   Is Clover the only way to do this?  Nothing else?  

Thanks.

**Comment by ArJoJo** [comment]
3x for your Q&A, I will try it and report back

**Comment by blekenbleu** [comment]
1. Get Bootdisk Utility from Softpedia [https://www.softpedia.com/get/System/Boot-Manager-Disk/Bootdisk-Utility.shtml#download](https://www.softpedia.com/get/System/Boot-Manager-Disk/Bootdisk-Utility.shtml#download)
2. Bootdisk Utility will install and configure Clover on a USB stick,
    but it does not configure the NVME driver:
    - copy NvmExpressDxe-64.efi from EFI > CLOVER > drivers > off to EFI > CLOVER > drivers > UEFI and EFI > CLOVER > drivers > BIOS
3. On my older Dell Inspiron, disabling Secure Boot and CSM (Load Legacy OPROM) worked.
4. I have yet to figure out how to remove useless options from Clover boot menu,
    but both bootable UEFI partitions are listed, first and second.

I documented this in [GitHub Pages](https://blekenbleu.github.io/NVMe/), but details change in different versions of BDU and Clover.
Perhaps other tools for booting Hackintosh can be repurposed to boot PCIe NVMe, but less easily...

**Reply by hafizullahchishti** [reply]
Still opaque.

I have a Samsung 980 (not Pro) SSD in the Sabrent adapter. The drive is visible from Windows and my partition utility. However, my mainboard UEFI does not see that drive as bootable and iterates it as a generic volume.

Using BDU to format the USB drive does not result in a bootable drive, nor can I view the contents from Windows GUI or cmd.exe subsequent to running BDU.

I can use Rufus to burn Clover-5151-X64.iso to the drive, and I boot to a Clover menu, but I can't tell where to go from there. 

I *did* copy NvmExpressDxe.efi to \EFI\CLOVER\drivers\BIOS, but that doesn't seem to give me a way to enumerate the NVMe drive in the Sabrent adapter.

The USB drive only boots in UEFI mode, not in legacy BIOS mode.

Please advise... Thanks, in advance.

**Reply by blekenbleu** [reply]
Are you trying to boot into Clover using the same drive that will run Windows?
My Clover boot drive is a separate USB thumb drive;  https://www.amazon.com/gp/product/B077Y149DL
I picked that thumb drive, because it barely protrudes from a USB socket and boots quickly.
Windows 10 on NVMe in PCIe adapter is a conventional UEFI installation.
That NVMe PCIe adapter wants to be in a slot with at least  4 lanes of PCIe,
which on most PC motherboards is a single slot occupied by a graphics adapter.
Documentation was updated to clarify that NVMe driver wanted copying to 2 locations.

**Reply by hafizullahchishti** [reply]
Hi, and thx for the reply.

My mainboard is ten years old and simply doesn't recognize the Samsung NVMe as bootable but just as a generic disk volume.

I have its adapter installed to a PCIeX16 slot.

What I'd like to do is put a boot record on the drive and clone Windows to it (possibly in a separate partition), but I'll settle for booting from a thumbdrive that passes bootstrap to Windows on the NVMe.

So my thumbdrive will boot into Clover, but there's no option to use it to install Windows to the NVMe drive. I *think* I've got the NvmExpressDxe.efi copied to the right directories, but the NVMe drive does not show in Clover's menu, nor do I have an option to run Windows' setup.exe from a separate source.

**Comment by blekenbleu** [comment]
Does your bootable Clover thumb drive show current bootable (non-NVMe) Windows partition?
If not, then Clover is not configured correctly on that thumb drive...
Clover finding boot partitions must work, to complete NVMe Windows installation.

Bootdisk Utility is now version 2.1.2022.030;
my success was on 21 Feb 2022;  NVMe boot configuration specifics may have changed.
Consider using Release 5144 from here:  https://github.com/CloverHackyColor/CloverBootloader/releases

Last time I tried, Windows 10 installation media from https://www.microsoft.com/en-us/software-download/windows10
when installed and booted from a thumb drive could install on NVMe in PCIe adapter.
To avoid accidently wiping out your current Windows installation, be sure to unplug that drive
before booting the Windows 10 installation media USB thumb drive.

**Comment by hafizullahchishti** [comment]
I was able to get my 10-year-old Asus M5A97_R2 mainboard to see my Samsung 980 NVMe drive using a hack I ran across on YouTube. Basically, it involves patching the UEFI with a generic NVMe driver. I've been running this for four days and it seems well-behaved and stable.

You will need the UEFI/BIOS binary for your board, UEFITool.exe (the utility to patch it), and NvmExpressDxe_4.ffs, the driver.

The video could be clearer, both visually, and logically, and it took me a couple of tries to get it to work.

In short, open the UEFI/BIOS file with UEFITool, do a text search for 
DXE
and choose an instance. Then look for "Volume Free Space." Go up one line, select it, and from the menu, choose "Insert After." The titlebar of the dialog should say "Select FFS file to insert". Use the dialog to navigate to where NvmExpressDxe_4.ffs is saved, and insert it.

Save your UEFI/BIOS file to a small USB stick freshly-formatted as FAT32. Power-down your system but do not remove power (i.e., unplug it).  Insert the USB stick into the USB "flashback" port, and press the "flash" button. On my board, it's inside, next to the IO sockets, but yours might be adjacent to the external port. You must use the flashback button; tryna update the UEFI any other way will probably not work and/or raise an error. 

Let it whir and grind until the indicator light goes dark or quits blinking, whichever your manual says to do.

Restart and you should find your NVMe driver iterated as a boot device in UEFI. You should be able to use a OS migrate-to-SSD utility to clone your existing Windows installation and boot from your NVMe drive.

The video is here... hope this helps. 
https://youtu.be/xFFBHkeosUU

**Comment by pterotronic** [comment]
installed clover on a usb. copied the driver to the specified directory. When I boot and press f10, I get the clover menu. All my boot drives show up, including the new nvme SSD w windows 10 on it. However, when I select it, I just get black screen and nothing happens. I'm sure it's not the hardware, it must be the clover file. when I boot normally, I can access the nvme ssd. However, when I'm at the clover menu, I still get the black screen even if I choose the normal sata SSD that the OS is on. So basically even though I get the clover menu, you can't boot to anything on it, even though it can see all the drives. How to fix?

**Reply by Aeroxia** [reply]
I'm having a similar issue. In the options that show up, 2 of 3 have the same name (there's only one install currently, windows 7), and all of them go to a black screen with flashing _ in the upper left, and won't reboot with ctrl + alt + del or power off normally, have to hold the power button to force it off. When I hit F3 (I think it was, from the F1 help), the USB stick options appear, and I was able to get past windows 10 saying it couldn't install on the NVMe in PCI-e adapter, but after it does some copying and gets to the privacy question quiz, it just reboots mid way with no error or explanation, so I can't proceed. I also can't boot my previous windows from Clover.

**Comment by alexandervlpl** [comment]
I prepared the Clover USB with Rufus (default settings). Copied the NVMe driver to both BIOS and UEFI folders. My Asus motherboard is in UEFI mode. In Clover I get about 5 mysterious boot options but the very first one is my Windows 10 on the NVMe, and it just works! :partying_face:

**Comment by pterotronic** [comment]
thanks for the info.
I'd kind of given up, because I never got any help that worked. But maybe I'll revisit and google up this rufus (don't know anything about it). 

very frustrating when windows sees the drive, but you can't boot from it. 

d

    On Wednesday, September 4, 2024 at 09:53:07 PM PDT, alexandervlpl ***@***.***> wrote:  
 
 

I prepared the Clover USB with Rufus (default settings). Copied theNVMe driver to both BIOS and UEFI folders. In Clover I get about 5 mysterious boot options but the very first one is my Windows 10 on the NVMe, and it just works! 🥳

**Comment by deepvision7** [comment]
Hi

Similar situation with hardware:
- An old MB with UEFI but without native NVMe support
- A Samsung M.2 disk
- An M.2 to PCIe adapter card
- An MBR active primary partition on another SSD where Windows 7 installed in legacy mode. 

I created an VHD image out of this system disk and put it on the NVMe .
Also added a new boot entry in the system BCD to boot from this VHD image.
But the Windows boot manager is not able to see the NVMe drive during boot.
So I was hoping to make Windows 7 boot manager to boot Windows from this VHD image with the help of Clover.

When I boot from the Clover thumbdrive it allows me to install Windows 10 from another thumbdrive to the NVMe drive.
But it won't see the existing Windows 7 installation and there is no option to boot into it.

Is there any way to make Clover see my old Windows 7 system installed in Legacy mode and allow to boot it with the NVMe support?

Thank in advance!

**Comment by hafizullahchishti** [comment]
You will need to “migrate” your existing installation to your boot drive to make the boot drive’s boot-record chain to the installed system.

There are utilities for this, and some of the free versions of partition managers and backup utilities such as from Macrorit and AOMEI have this built-in.

**Reply by deepvision7** [reply]
Thanks. I was hoping to leave the existing system with the boot record intact for now just in case. And since the Clover is able to see the NVMe drive maybe there is a way to make it boot the VHD image from it somehow.

**Comment by hafizullahchishti** [comment]
“I was hoping to leave the existing system with the boot record intact for now”


That’s what I did. I can still boot from Drive 1 if I wish, by choosing it in the UEFI. So you’re not actually moving the old bootable system from the drive, just cloning it.

**Comment by deepvision7** [comment]
Not sure I understand how to do this without cloning the entire system to the new drive. Or that's exactly what I should do? Thanks.

**Comment by hafizullahchishti** [comment]
Yes. Clone the current boot drive to the new one, and set the new one as the boot device in UEFI. You may have to also disable Secure Boot, which is another setting.

If after that UEFI does not see the cloned drive as a boot device, you may have to patch the UEFI, or use a utility to start the bootstrap from Clover and chain to the current boot drive.

**Reply by deepvision7** [reply]
Well, I can install the new Windows 10 to NVMe in UEFI, it works perfectly with Clover. Don't need the old system on the new disk. I was just hoping to make Clover also boot into VHD image with my old system on the NVMe drive somehow.

**Reply by deepvision7** [reply]
Actually made it to work by booting Clover in UEFI, then loading Grub2, and then choosing VHD file to boot.

**Comment by pterotronic** [comment]
I gave up on this. Nothing seems to work.
Originally, I would boot up and see the drive on the expansion card. It just wasn't visible in the boot section of my bios.
However, probably due to Microsoft update, I don't even see the drive when I boot up anymore and it's now absent even from device driver. 
I'm not willing to flash my bios. If there is any chance this would brick my machine, it's just not worth that kind of risk. 

d
    On Sunday, February 16, 2025 at 12:44:36 AM PST, deepvision7 ***@***.***> wrote:  
 
 


Hi

Similar situation with hardware:
   
   - An old MB with UEFI but without native NVMe support
   - A Samsung M.2 disk
   - An M.2 to PCIe adapter card
   - An MBR active primary partition on another SSD where Windows 7 installed.

I created an VHD image out of this system disk and put it on the NVMe .
Also added a new boot entry in the system BCD to boot from this VHD image.
But the Windows boot manager is not able to see the NVMe drive during boot.
So I was hoping to make Windows 7 boot manager to boot Windows from this VHD image with the help of Clover.

When I boot from the Clover thumbdrive it allows me to install Windows 10 from another thumbdrive to the NVMe drive.
But it won't see the existing Windows 7 installation and there is no option to boot into it.

Is there any way to make Clover see my old Windows 7 system and allow to boot it with the NVMe support?

Thank in advance!

**Comment by hafizullahchishti** [comment]
I believe that Clover will allow you to boot from a very small USB storage device and then chain the bootstrap sequence to your NVMe drive even though your BIOS/UEFI does support booting from NVMe. I toyed with this and even bought a very small USB drive for the purpose, but found a patch for my UEFI that lets it see my Samsung NVMe drive.



I wish I could help you, but I never followed-through with Clover for that purpose.

**Comment by hafizullahchishti** [comment]
Yeah, I think that was the sequence on the Clover website. Glad you got it to work!

 

And you have an additional security feature: take the thumbdrive with you and your system won’t boot without it.
